### ISSUE TITLE
consistent Point type

### ISSUE THREAD

**Issue by wheybags** [wheybags/freeablo#326]
We have a million different ways of representing a pair of int32_ts, sometimes we use std::pair<int32_t>, sometimes we pass separate x and y values, and there's a few places where we've even made structs with an x and y member.
It would be nice to create a single Point type (probably in the misc component), and just use that consistently. Note that the Position class cannot be used for this, as it represents more than just a single point (it represents two points side by side and an interpolation between them)

**Comment by Hydroque** [comment]
Should it be a definite use of std::pair<T, V>, I mean it is well defined.

**Comment by wheybags** [comment]
I'd prefer a dedicated class with an X and y member. That way we don't need
to write first and second all the time, and we can add member functions

**Comment by Hydroque** [comment]
Correct, but standard pair is what we are looking for and then and if we need more we can abstract it. I mean we can scrap the std pair if you think it doesn't fit. It's an easy fix

**Comment by Hydroque** [comment]
I am thinking of something like this, it's a header only

__pair.hpp__
```

#pragma once

namespace Misc
{
	template<typename T, typename V> class TPair
	{
	public:
		T x;
		V y;
		inline TPair(T x, V y)
		{
			this->x = x;
			this->y = y;
		}
	};
	
	template<typename T> class SPair : public TPair<T, T>
	{
	public:
		inline SPair(T x, T y)
			: TPair<T, T>(x, y)
		{
			
		}
	};
}

```

__Usage:__
```
#include <iostream>

#include "pair.hpp"

int main() {
	
	Misc::TPair<int32_t, int32_t> arr1(1, 2);
	Misc::SPair<int32_t> arr2(3, 4);
	
	std::cout << "first: " << arr1.x << ", " << arr1.y << std::endl;
	std::cout << "second: " << arr2.x << ", " << arr2.y << std::endl;
	
	return 0;
}
```

__Which should be the rough equivalent of https://godbolt.org/g/MNWMzz__
```
int32_t x1;
int32_t y1;
x1 = 1;
y1 = 2;

int32_t x2;
x2 = 3;
y2 = 4;
```

**Comment by Predelnik** [comment]
What's the point of having templated pair when for point you want fixed type and definetily not different types for x/y? As for your `SPair` what's the cardinal difference with it and `std::array<int32_t, 2>`?
Simplest implementation of point ( + maybe the same with doubles):

    struct point
    {
        int32_t x;
        int32_t y;
    };

If you want comparison (which is by the way very rare thing for points) write it using `std::tie`. More useful thing for points is arithmetic operations because point could also be used as mathematical vector and it would be a bit clumsy to attach these operations to `pair`.

Possible point class example http://doc.qt.io/archives/qt-4.8/qpoint.html

**Comment by Hydroque** [comment]
Youch it was just an example of something I had no direction over.

**Comment by Predelnik** [comment]
Well I just wanted to point out why it's better to avoid pair in this case. In general pair usage should be very rare and even the fact that it's used for `std::map` is rather unfortunate.

**Comment by Hydroque** [comment]
Its rather cheap to have a generic/template because you never know what you will need a pair for. Saves rewriting code. For what magnitude it was to be used is unknownst and I am just trying to help here. The information given wasn't sufficient and I shouldn't have taken up the issue, but I really wrote the code as a question.

**Comment by Predelnik** [comment]
Don't worry, I don't even know the scope of point usage precisely myself, but I couldn't help to notice that `pair` is really overused in freeablo code. Yes `pair` is generic but it doesn't express anything that's why it makes code less readable and less maintanable.
